src/aliceVision/matching/RegionsMatcher.hpp

```diff
@@ -145,7 +145,7 @@ class RegionsMatcher : public IRegionsMatcher
     if (!matcher_.SearchNeighbours(queries, queryregions_.RegionCount(), &vec_nIndice, &vec_fDistance, NNN__))
       return false;
 
-    assert(vec_nIndice.size() == vec_fDistance.size());
+    //assert(vec_nIndice.size() == vec_fDistance.size());
```

Comment:
So once I added OpenMP, this line would fail, but when I commented it out the sample code worked fine as before. Is this required, shouldn't we only be doing assertions in the unit test?

Reply:
@rjanvier , could it be possible the assertion is failing due to a race condition?

Reply:
Sure it is :). you are missing an "OpenMP critical section" before the result loop [see here](https://github.com/openMVG/openMVG/blob/cf5a5754926b7e73c9e1d0f67fac296db2365e5f/src/openMVG/matching/matcher_hnsw.hpp#L147-L149)

this make me think you can also try direct indexing, in theory it's far better, but in practice I bet you won't notice the difference. Something like that:
```
  pvec_indices->resize(nbQuery * NN);
  pvec_distances->resize(nbQuery * NN);
...
...
    #Your current OpenMP pragma
    for (int i = 0; i < nbQuery; ++i) {
      const auto result = HNSWmatcher->searchKnn((const void *) (query + dimension_ * i), NN,
        [](const std::pair<DistanceType, size_t> &a, const std::pair<DistanceType, size_t> &b) -> bool {
          return a.first < b.first;
      });
      for (int j = 0; j < NN; ++j)
      {
        const auto & res = result[j];
        (*pvec_indices)[i*NN+j] = IndMatch(i, res.second); ;
        (*pvec_distances)[i*NN+j] = res.first;
      }
    }```

Reply:
Ah my bad, silly mistake to forget that critical section. I have added it, and I removed the commenting out of that one assertion statement in `RegionsMatcher.hpp`, and it doesn't fail anymore; everythings works great!

Reply:
@rjanvier Awesome, thank you for the direct indexing solution, about to push it soon. I was just curious, other than avoiding local variable issues during multi-threading, what are the advantages of the direct indexing solution?

Reply:
In theory, by using "critical section", only one thread can do the update at time so it's possible you loose performances somehow (because some threads could be waiting their turn to emplace in the vector)... 
But In theory too, `vector::resize` should have an overhead vs. `vector::reserve + emplace_back`, so it's a mixed bag and it depends also on how compilers optimize this. On my side, with a  micro becnhmark of 10 000k pairs of images, I noticed the two approaches are surprisingly very similar in term of execution speed... but I do prefer the semantic of "direct indexing" (and I think everybody should)

Reply:
That makes sense, with very similar performance a static approach rather than multithreading seems safer. I also just pushed the direct indexing approach, please let me know if anything could be improved or changed.

Reply:
I'm not a software engineer and I think you have more clever and skilled people than me at your disposal. I'm confident they can help you to sort that out :)